### PR TITLE
TS Migration

### PR DESCRIPTION
**Pull request by ersbeth**

### CODE REVIEW

File: src/blobtree/Element.ts [hunk 1]
```diff
@@ -30,14 +30,11 @@ let elementIds = 0;
...
-export class Element {
+export abstract class Element {
 
     static type = "Element";
 
-    /**
-     * @param {ElementJSON} _json
-     */
-    static fromJSON(_json: any) {
+    static fromJSON(_json: ElementJSON) {
```
Comment (ersbeth) [inline src/blobtree/Element.ts:40]: Can't static methods be abstract ?

File: src/blobtree/Element.ts [hunk 2]
```diff
@@ -244,11 +233,9 @@ export class Element {
...
 
-    destroy() {
-        console.error("Blobtree.Element: destroy is a virtual function, should be reimplemented in all classes extending Element.");
-    }
+    abstract destroy(): void;
 
 };
 
-Types.register(Element.type, Element);
+Types.register(Element.type,  {fromJSON: Element.fromJSON});
```
Comment (ersbeth) [inline src/blobtree/Element.ts]: Why don't you register the whole type ?

File: src/blobtree/Material.ts [hunk 3]
```diff
@@ -45,9 +39,9 @@ export class Material {
      *  @param {Array.<Material>=} arr4
      *  @param {Array.<Material>=} arr5
      *
-     *  @return {boolean} true if and only if all arguments are arrays of the same length and containing the same material values.
+     *  @return true if and only if all arguments are arrays of the same length and containing the same material values.
      */
-    static areEqualsArrays(arr1) {
+    static areEqualsArrays(arr1: Material[]): boolean {
 
         console.warn("Material.areEqualsArrays is deprecated, please use your own comparison function using Material.equals.");
```
Comment (ersbeth) [inline src/blobtree/Material.ts:45]: Add a @deprecated tag in the JSDOC

File: src/blobtree/Material.ts [hunk 4]
```diff
@@ -157,33 +151,29 @@ export class Material {
...
-    getColor() { return this.color; };
+    getColor(): Color { return this.color; };
 
-    /** @return {number} */
-    getRoughness() { return this.roughness; };
+    getRoughness(): number { return this.roughness; };
 
-    /** @return {number} */
-    getMetalness = function () { return this.metalness; };
+    getMetalness = function (this: Material): number { return this.metalness; };
```
Comment (ersbeth) [inline src/blobtree/Material.ts]: shouldn't this be defined like the other accessors ? ie. `getMetalness(){}` instead of `getMetalness = function`. Second one is not included in prototype but in a field.
With the first one you shouldn't need to type `this` in the signature.

File: src/blobtree/MaxNode.ts [hunk 5]
```diff
@@ -19,59 +14,51 @@ import { Material } from "./Material";
...
+            const self = this;
             children.forEach(function (c) {
                 self.addChild(c);
             });
         }
 
-        // temp vars to speed up evaluation by avoiding allocations
-        /** @type {{v:number, g:Vector3, m:Material}} */
+        // temp consts to speed up evaluation by avoiding allocations
         this.tmp_res = { v: 0, g: null, m: null };
```
Comment (ersbeth) [inline src/blobtree/MaxNode.ts]: These init values don't depend on constructor parameters. Thus they should be directly initialized at declaration above. Same for the following lines. In general prefer this pattern when possible: avoid initialization in constructor when it can be made at declaration.

File: src/blobtree/MinNode.ts [hunk 6]
```diff
@@ -20,58 +14,53 @@ import { Material } from "./Material";
...
+            const self = this;
             children.forEach(function (c) {
                 self.addChild(c);
             });
         }
 
-        // temp vars to speed up evaluation by avoiding allocations
-        /** @type {{v:number, g:Vector3, m:Material}} */
+        // temp consts to speed up evaluation by avoiding allocations
         this.tmp_res = { v: 0, g: null, m: null };
```
Comment (ersbeth) [inline src/blobtree/MinNode.ts]: Same as above

File: src/blobtree/Node.ts [hunk 7]
```diff
@@ -55,25 +46,22 @@ export class Node extends Element {
     /**
      *  Clone current node and itss hierarchy
```
Comment (ersbeth) [inline src/blobtree/Node.ts]: fix typo in comment

File: src/blobtree/Node.ts [hunk 8]
```diff
@@ -248,5 +227,5 @@ export class Node extends Element {
 
 };
 
-Types.register(Node.type, Node);
+Types.register(Node.type, {fromJSON: Node.fromJSON});
```
Comment (ersbeth) [inline src/blobtree/Node.ts]: Same as previous, why don't you register the whole class?

File: src/blobtree/Primitive.ts [hunk 9]
```diff
@@ -85,29 +76,27 @@ export class Primitive extends Element {
...
-    getAreas(): { aabb: THREE.Box3, bv: Area, obj: Primitive }[] {
+    override getAreas(): { aabb: THREE.Box3, bv: Area, obj: Primitive }[] {
         console.error("ERROR : getAreas is an abstract function, should be re-implemented in all primitives(error occured in " + this.getType() + " primitive)");
         return [];
     };
 
     /**
      * @abstract
-     * Compute variables to help with value computation.
+     * Compute constiables to help with value computation.
```
Comment (ersbeth) [inline src/blobtree/Primitive.ts]: This is an akward naming, let variables instead, it's ok even if they don't truly vary.

File: src/blobtree/Primitive.ts [hunk 10]
```diff
@@ -85,29 +76,27 @@ export class Primitive extends Element {
...
      */
-    computeHelpVariables() {
-        throw "ERROR : computeHelpVariables is a virtual function, should be re-implemented in all primitives(error occured in " + this.getType() + " primitive)";
-    };
+    abstract override computeHelpVariables(): void;
 
     /**
      * @abstract
-     * Compute variables to help with value computation.
+     * Compute constiables to help with value computation.
```
Comment (ersbeth) [inline src/blobtree/Primitive.ts]: same as above

File: src/blobtree/RicciNode.ts [hunk 11]
```diff
@@ -39,34 +41,27 @@ export class RicciNode extends Node {
             });
         }
 
-        // Tmp vars to speed up computation (no reallocations)
-        /** @type {Float32Array} */
+        // Tmp consts to speed up computation (no reallocations)
         this.tmp_v_arr = new Float32Array(0);
```
Comment (ersbeth) [inline src/blobtree/RicciNode.ts]: same as before, init at declaration when possible.

File: src/blobtree/RootNode.ts [hunk 12]
```diff
@@ -384,7 +361,7 @@ export class RootNode extends RicciNode {
                         point: curPos.clone(),
                         g: dicho_res.g.clone()
                     });
-                    // set variable in order to resume to where we were
+                    // set constiable in order to resume to where we were
```
Comment (ersbeth) [inline src/blobtree/RootNode.ts]: idem leave variable

File: src/blobtree/ScaleNode.ts [hunk 13]
```diff
@@ -20,29 +20,30 @@ import { Material } from "./Material";
...
+            const self = this;
             children.forEach(function (c) {
                 self.addChild(c);
             });
         }
 
-        // temp vars to speed up evaluation by avoiding allocations
-        /** @type {{v:number, g:Vector3, m:Material}} */
+        // temp consts to speed up evaluation by avoiding allocations
         this.tmp_res = { v: 0, g: null, m: null };
```
Comment (ersbeth) [inline src/blobtree/ScaleNode.ts]: idem init at declaration

File: src/blobtree/areas/AreaCapsule.ts [hunk 14]
```diff
@@ -93,24 +113,21 @@ export class AreaCapsule extends Area {
...
+                const tmp = sphere.radius + inter_w;
                 return (dist < tmp * tmp);
             }
         }
     }
 
     /**
      * @link Area.contains for a complete description
-     * @param {Vector3} p
+     * @param p
```
Comment (ersbeth) [inline src/blobtree/areas/AreaCapsule.ts]: remove this

File: src/blobtree/areas/AreaScalisSeg.ts [hunk 15]
```diff
@@ -128,133 +143,99 @@ export class AreaScalisSeg extends Area {
...
+     *  @param sphere  A sphere object, must define sphere.radius (radius) and sphere.center (center, as a Vector3)
+     *  @param factor  the ratio to determine the wanted accuracy.
      *
+     *  @return the accuracy needed in the intersection zone
      *  @todo Check the Maths
      */
-    getAcc(sphere, factor) {
+    getAcc(sphere: AreaSphereParam, factor: number): number {
         this.proj_computation(sphere.center);
-        /*
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisSeg.ts:153]: You should not delete this stuff unless @maximeq told you to do so. If he didnt, ask him if he wants to keep this or not.

File: src/blobtree/areas/AreaScalisSeg.ts [hunk 16]
```diff
@@ -265,27 +246,25 @@ export class AreaScalisSeg extends Area {
             thick1 = this.thick0;
         }
 
-        var diff = t - p0[axis];
+        let diff = t - p0[axis];
         if (diff < -2 * thick0) {
             step = Math.min(step, Math.max(Math.abs(diff + 2 * thick0), Accuracies.curr * thick0));
         } else if (diff < 2 * thick0) {
             step = Math.min(step, Accuracies.curr * thick0);
-        }// else the vertex is behind us
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisSeg.ts:273]: keep those comments, this is documentation

File: src/blobtree/areas/AreaScalisSeg.ts [hunk 17]
```diff
@@ -265,27 +246,25 @@ export class AreaScalisSeg extends Area {
...
+        }
         diff = t - p1[axis];
         if (diff < -2 * thick1) {
             step = Math.min(step, Math.max(Math.abs(diff + 2 * thick1), Accuracies.curr * thick1));
         } else if (diff < 2 * thick1) {
             step = Math.min(step, Accuracies.curr * thick1);
-        }// else the vertex is behind us
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisSeg.ts:279]: same

File: src/blobtree/areas/AreaScalisTri.ts [hunk 18]
```diff
@@ -1,35 +1,60 @@
...
+        v: [ScalisVertex, ScalisVertex, ScalisVertex],
+        unit_normal: Vector3,
+        main_dir: Vector3,
+        segParams: any,
+        min_thick: number,
+        max_thick: number
+    ) {
+        super();
 
         this.tmpVect = new Vector3();
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisTri.ts]: init at declaration when possible

File: src/blobtree/areas/AreaScalisTri.ts [hunk 19]
```diff
@@ -53,79 +79,110 @@ export class AreaScalisTri extends Area {
...
      *  Compute projection (used in other functions)
-     *  @param {!Vector3} p Point to proj
-     *  @param {!Object} segParams A seg param object @todo clarify this parameter
+     *  @param p Point to proj
+     *  @param segParams A seg param object
      *
      *  @protected
      */
-    proj_computation(p, segParams) {
+    protected proj_computation(p: Vector3, segParams: any): void {
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisTri.ts]: investigate the correct type for segParams, don't use any

File: src/blobtree/areas/AreaScalisTri.ts [hunk 20]
```diff
@@ -137,23 +194,25 @@ export class AreaScalisTri extends Area {
     /**
      * @link Area.sphereIntersect for a complete description
      * @todo Check the Maths (Ask Cedric Zanni?)
-     * @param {AreaSphereParam} sphere
-     * @return {boolean} true if the sphere and the area intersect
+     * @param sphere
+     * @
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisTri.ts]: problem with spacing

File: src/blobtree/areas/AreaScalisTri.ts [hunk 21]
```diff
@@ -137,23 +194,25 @@ export class AreaScalisTri extends Area {
...
-            let intersectSeg = this.sphereIntersectSegment(sphere, this.segParams[i], ScalisMath.KS);
+            const intersectSeg = this.sphereIntersectSegment(sphere, this.segParams[i], ScalisMath.KS);
             // The sphere intersecting ones the angle means the sphere intersect the Bounding Volume
             if (intersectSeg) { return true; }
         }
         // Second: Test the intersection of the sphere with the triangular prism defined by
         // the 2D triangle constructed from the vertices and of half heights Ti*KS along the unit_normal for each vertices Vi
         let inside = true;
-        for (let i = 0, inside = true; i < 5; i++) {
+        for (let i = 0; i < 5; i++) {
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisTri.ts:208]: maybe you should keep `inside` in the loop declaration if you don't need it outside

File: src/blobtree/areas/AreaScalisTri.ts [hunk 22]
```diff
@@ -297,102 +348,99 @@ export class AreaScalisTri extends Area {
...
      */
-    getRawAcc(sphere) {
+    getRawAcc(sphere: AreaSphereParam): number {
         return this.getAcc(sphere, Accuracies.raw);
     }
 
     /**
      * @link Area.getMinAcc
-     * @return {number}
+     * @return number
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisTri.ts]: remove this

File: src/blobtree/scalis/ScalisPoint.ts [hunk 23]
```diff
@@ -55,37 +47,37 @@ export class ScalisPoint extends ScalisPrimitive {
         this.materials.push(mat);
     }
 
-    getType() {
+    override getType(): string {
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisPoint.ts]: return type should be a literal, not string

File: src/blobtree/scalis/ScalisSegment.ts [hunk 24]
```diff
@@ -100,76 +100,76 @@ export class ScalisSegment extends ScalisPrimitive {
...
-    // [Abstract] See Primitive.setVolType for more details
-    setVolType(vt: ScalisPrimitiveVolType) {
+    /**
+     *  [Abstract] See Primitive.setVolType for more details.
+     *  @param vt New VolType to set (Only for SCALIS primitives)
+     */
+    override setVolType(vt: ScalisPrimitiveVolType): void {
         if (!(vt == ScalisPrimitive.CONVOL || vt == ScalisPrimitive.DIST)) {
-            throw "ERROR : volType must be set to ScalisPrimitive.CONVOL or ScalisPrimitive.DIST";
+            throw new Error("volType must be set to ScalisPrimitive.CONVOL or ScalisPrimitive.DIST");
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisSegment.ts]: [ScalisSegment] setVolType: volType must be set to ScalisPrimitive.CONVOL or ScalisPrimitive.DIST

File: src/blobtree/scalis/ScalisSegment.ts [hunk 25]
```diff
@@ -100,76 +100,76 @@ export class ScalisSegment extends ScalisPrimitive {
...
     // [Abstract] See Primtive.getArea for more details
-    getAreas(): {
+    override getAreas(): {
         aabb: Box3,
         bv: AreaScalisSeg,
         obj: ScalisSegment,
     }[] {
         if (!this.valid_aabb) {
-            console.error("ERROR : Cannot get area of invalid primitive");
+            console.error("ERROR: Cannot get area of invalid primitive");
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisSegment.ts]: same as above, format error

File: src/blobtree/scalis/ScalisSegment.ts [hunk 26]
```diff
@@ -197,27 +197,26 @@ export class ScalisSegment extends ScalisPrimitive {
...
 
-        // Speed up var for cylinder bounding
+        // Speed up const for cylinder bounding
         // Used only in evalConvol
         this.cyl_bd0 = Math.min(-bound_supp0, this.length - bound_supp1);
         this.cyl_bd1 = Math.max(this.length + bound_supp1, bound_supp0);
 
         this.increase_unit_dir.copy(this.unit_dir);
-        // weight help variables
+        // weight help constiables
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisSegment.ts]: variables

File: src/blobtree/scalis/ScalisSegment.ts [hunk 27]
```diff
@@ -238,9 +237,9 @@ export class ScalisSegment extends ScalisPrimitive {
                 this.evalConvol(p, res);
                 break;
             default:
-                throw "Unknown volType, cannot evaluate.";
+                throw new Error("Unknown volType, cannot evaluate.");
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisSegment.ts]: format error

File: src/blobtree/scalis/ScalisSegment.ts [hunk 28]
```diff
@@ -344,81 +344,79 @@ export class ScalisSegment extends ScalisPrimitive {
...
     ///////////////////////////////////////////////////////////////////////////
     // Convolution Evaluation functions and auxiliaary functions
     /**
      *  value function for Convol volume type (Homothetic convolution).
      */
-    evalConvol(p: Vector3, res: ValueResultType) {
+    evalConvol(p: Vector3, res: ValueResultType): void {
         if (!this.valid_aabb) {
-            throw "Error : prepareForEval should have been called";
+            throw new Error("prepareForEval should have been called");
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisSegment.ts]: format error

File: src/blobtree/scalis/ScalisTriangle.ts [hunk 29]
```diff
@@ -549,7 +561,7 @@ export class ScalisTriangle extends ScalisPrimitive {
         point: Vector3,
         p1: Vector3,
         p1p2: Vector3,
-        length: number,
+        _length: number, // Unused parameter
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisTriangle.ts]: remove if this is really not used

File: src/blobtree/scalis/ScalisTriangle.ts [hunk 30]
```diff
@@ -662,17 +678,23 @@ export class ScalisTriangle extends ScalisPrimitive {
                 res.v *= factor;
                 if (res.g) {
                     var grad_res = new Vector3();
-                    grad_res.addVectors(grad_res, res_low.g);
+                    const res_low_g = res_low.g;
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisTriangle.ts]: why did you introduce variables ?

File: src/blobtree/scalis/ScalisTriangle.ts [hunk 31]
```diff
@@ -921,4 +945,4 @@ export class ScalisTriangle extends ScalisPrimitive {
     }
 };
 
-Types.register(ScalisTriangle.type, ScalisTriangle);
\ No newline at end of file
+Types.register(ScalisTriangle.type,  {fromJSON: ScalisTriangle.fromJSON});
```
Comment (ersbeth) [inline src/blobtree/scalis/ScalisTriangle.ts]: why don't you register the class ?

File: src/blobtree/sdf/Poly6DistanceFunctor.ts [hunk 32]
```diff
@@ -1,37 +1,31 @@
...
     }
 
     /**
      * This is the standard 6 degree polynomial function used for implicit modeling.
      * At 0, its value is 1 with a zero derivative.
      * At 1, its value is 0 with a zero derivative.
-     * @param {number} d
      */
-    static evalStandard(d) {
+    evalStandard(d: number): number {
```
Comment (ersbeth) [inline src/blobtree/sdf/Poly6DistanceFunctor.ts]: why did you remove the `static` ? It is possible that an app somewhere calls this from the class and not from an instance.

File: src/blobtree/sdf/Poly6DistanceFunctor.ts [hunk 33]
```diff
@@ -67,34 +58,31 @@ export class Poly6DistanceFunctor extends DistanceFunctor {
...
+     * @returns dimensional gradient at d.
      */
-    gradient(d) {
-        var ds = d / (2 * this.scale) + 0.5;
-        var res = (1 - ds * ds);
-        res = -(6 / (2 * this.scale)) * ds * res * res / Poly6DistanceFunctor.evalStandard(0.5);
+    override gradient(d: number): number {
+        const ds = d / (2 * this.scale) + 0.5;
+        let res = 1 - ds * ds;
+        res = -(6 / (2 * this.scale)) * ds * res * res / this.evalStandard(0.5);
```
Comment (ersbeth) [inline src/blobtree/sdf/Poly6DistanceFunctor.ts]: Related to previous one

File: src/blobtree/sdf/SDFCapsule.ts [hunk 34]
```diff
@@ -1,25 +1,23 @@
...
  *  You can find more on Capsule geometry here https://github.com/maximeq/three-js-capsule-geometry
  *
  *  @constructor
  *  @extends SDFPrimitive
- *
  */
 export class SDFCapsule extends SDFPrimitive {
 
-    static type = "SDFCapsule" as const;
+    static override type = "SDFCapsule";
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFCapsule.ts]: keep const as it narrows the type

File: src/blobtree/sdf/SDFNode.ts [hunk 35]
```diff
@@ -1,89 +1,79 @@
...
+            new Vector3(+Infinity, +Infinity, +Infinity)
         );
 
-        /** @type {Array<SDFNode|SDFPrimitive>} */
-        this.children;
+        this.children = [];
     }
 
-    getType() {
+    overridegetType(): string {
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFNode.ts]: hum there should be a syntax error here

File: src/blobtree/sdf/SDFNode.ts [hunk 36]
```diff
@@ -1,89 +1,79 @@
...
     /**
-     *  SDF Field are infinite, so Areas do not make sens except for the SDFRoot, which will
+     *  SDF Field are infinite, so Areas do not make sense except for the SDFRoot, which will
      *  usually apply a compact kernel to the distance field.
      *  @abstract
-     *  @return {Object}
      */
-    getAreas() {
+    override getAreas(): { aabb: Box3; bv: Area; obj: Primitive; }[] {
         throw "No Areas for SDFNode, except for the SDFRootNode.";
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFNode.ts]: format error

File: src/blobtree/sdf/SDFNode.ts [hunk 37]
```diff
@@ -95,22 +85,23 @@ export class SDFNode extends Node {
...
     };
 
-    // Abstract
-    /**
-     * @abstract
-     * @return {number}
-     */
-    heuristicStepWithin() {
+    override heuristicStepWithin(): number {
         throw "heuristicStepWithin may not make sens for all SDFNode, except for the SDFRootNode.";
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFNode.ts]: format error

File: src/blobtree/sdf/SDFPoint.ts [hunk 38]
```diff
@@ -93,19 +93,14 @@ export class SDFPoint extends SDFPrimitive {
     /**
      * @link SDFPrimitive.getDistanceAreas
      * @param d Distance to consider for the area computation.
-     * @returns {Array.<>}
      */
     getDistanceAreas(d: number): { aabb: Box3, bv: Area, obj: SDFPoint }[] {
         if (!this.valid_aabb) {
-            throw "ERROR : Cannot get area of invalid primitive";
+            throw new Error("ERROR : Cannot get area of invalid primitive");
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFPoint.ts]: format error

File: src/blobtree/sdf/SDFPoint.ts [hunk 39]
```diff
@@ -115,24 +110,21 @@ export class SDFPoint extends SDFPrimitive {
...
      */
     value = (function () {
-        var v = new Vector3();
+        const v = new Vector3();
 
-        return function (p: Vector3, res: ValueResultType) {
+        return function (this: SDFPoint, p: Vector3, res: ValueResultType): void {
             if (!this.valid_aabb) {
-                throw "Error : PrepareForEval should have been called";
+                throw new Error("Error : PrepareForEval should have been called");
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFPoint.ts]: format error

File: src/blobtree/sdf/SDFRootNode.ts [hunk 40]
```diff
@@ -1,189 +1,132 @@
...
 
         this.f = f;
-
         this.materials.push(material ? material.clone() : new Material());
+        this.sdfRoot = sdfRoot ? (sdfRoot instanceof SDFNode ? sdfRoot : new SDFNode().addChild(sdfRoot)) : new SDFNode();
 
-        this.sdfRoot = sdfRoot ?
-            (sdfRoot instanceof SDFNode ? sdfRoot : new SDFNode().addChild(sdfRoot)) : new SDFNode();
-
-        // Tmp vars to speed up computation (no reallocations)
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFRootNode.ts:70]: keep those comments

File: src/blobtree/sdf/SDFRootNode.ts [hunk 41]
```diff
@@ -1,189 +1,132 @@
...
-    /**
-     * @param {SDFNode | SDFPrimitive} c
-     */
-    addChild(c) {
+    addChild(c: SDFNode | SDFPrimitive): void {
         if (this.sdfRoot.children.length === 0) {
             this.sdfRoot.addChild.call(this, c);
         } else {
-            throw "Error : SDFRootNode can have only one child.";
+            throw new Error("SDFRootNode can have only one child.");
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFRootNode.ts]: Format error

File: src/blobtree/sdf/SDFRootNode.ts [hunk 42]
```diff
@@ -1,189 +1,132 @@
...
             sdfRoot: this.sdfRoot.toJSON()
         };
-        return res;
-    };
+    }
 
-    prepareForEval() {
+    prepareForEval(): void {
         if (!this.valid_aabb) {
-            this.aabb = new Box3();  // Create empty BBox
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFRootNode.ts:112]: keep comments

File: src/blobtree/sdf/SDFRootNode.ts [hunk 43]
```diff
@@ -1,189 +1,132 @@
...
-
             this.valid_aabb = true;
         }
-    };
+    }
 
-    /**
-     *  @link Element.getAreas for a complete description
-     *
-     *  This function is an attempt to have SDFRootNode behave like a Primitive in the normal Blobtree.
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFRootNode.ts:128]: keep jsdoc

File: src/blobtree/sdf/SDFRootNode.ts [hunk 44]
```diff
@@ -1,189 +1,132 @@
...
+                aabb: area.aabb,
+                bv: area.bv,
+                obj: this
+            }));
         }
-    };
+    }
 
-    /**
-     *  @link Node.value for a complete description
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFRootNode.ts:150]: keep jsdoc

File: src/blobtree/sdf/SDFRootNode.ts [hunk 45]
```diff
@@ -1,189 +1,132 @@
...
             if (res.g) {
-                res.g.copy(tmp.g).multiplyScalar(this.f.gradient(res.v))
+                res.g.copy(tmp.g!).multiplyScalar(this.f.gradient(res.v));
             }
             if (res.m) {
                 res.m.copy(this.materials[0]);
             }
-        }
-        else if (res.step !== undefined) {
-            // return distance to aabb such that next time we'll hit from within the aabbb
```
Comment (ersbeth) [inline src/blobtree/sdf/SDFRootNode.ts:183]: keep comment

File: src/utils/Convergence.ts [hunk 46]
```diff
@@ -109,39 +99,7 @@ Convergence.safeNewton3D = function (pot,              // Scalar Field to eval
         res.copy(starting_point);
         return;
     }
-
-    /*
-    if(broken){
```
Comment (ersbeth) [inline src/utils/Convergence.ts:114]: Ask @maximeq if its ok to delete this

File: src/utils/TriangleUtils.ts [hunk 47]
```diff
@@ -42,11 +117,8 @@ export const TriangleUtils = {};
...
  * intermediary functions used in computeVectorsDirs
- * @param {number} ind
- * @param {number} lengthArray
- * @return {number}
  */
-let cleanIndex = function (ind, lengthArray) {
+cleanIndex(ind: number, lengthArray: number) {
     let res = ind;
     if (lengthArray === 0) {
         throw new Error("Lenght of the array should not be 0");
```
Comment (ersbeth) [inline src/utils/TriangleUtils.ts]: Format error

File: src/utils/TriangleUtils.ts [hunk 48]
```diff
@@ -61,56 +133,196 @@ let cleanIndex = function (ind, lengthArray) {
...
- * @property {Vector3} point_half
- * @property {Vector3} half_dir_2
- * @property {number} coord_max
- * @property {number} coord_middle
- * @property {number} unit_delta_weight
- * @property {Vector3} longest_dir_special
- * @property {number} max_seg_length = tmp.length();
- * @property {Vector3} unsigned_ortho_dir = triangle.ortho_dir.clone();
- */
+    triangle.p0p1 ? triangle.p0p1.subVectors(v1_p, v0_p) : null;
```
Comment (ersbeth) [inline src/utils/TriangleUtils.ts]: Strange expression: you don't store the result anywhere. If you are interested in side effect only you should use an instruction instead.

File: src/utils/TriangleUtils.ts [hunk 49]
```diff
@@ -61,56 +133,196 @@ let cleanIndex = function (ind, lengthArray) {
...
 
 /**
- * A number, or a string containing a number.
- * @typedef {Object} VertexLike
- * @property {() => Vector3} getPos
- * @property {() => number} getThickness
+ * Updates the cached values of the triangle
+ * @param triangle The triangles who's internal values need to be updated
  */
+updateComputedAttributes(triangle: TriangleLike) {
```
Comment (ersbeth) [inline src/utils/TriangleUtils.ts:142]: Where does this function come from ?
Diff seems to be broken here I cant review that

File: src/utils/TriangleUtils.ts [hunk 50]
```diff
@@ -170,7 +382,7 @@ TriangleUtils.computeVectorsDirs = function (triangle) {
             triangle.ortho_dir = dir_1.clone();
             triangle.ortho_dir.normalize();
 
-            // direction of fastest variation of weight
+            // direction of fastest constiation of weight
```
Comment (ersbeth) [inline src/utils/TriangleUtils.ts]: hum constiation does nt make sense at all for me.
keep variation

File: src/utils/TriangleUtils.ts [hunk 51]
```diff
@@ -185,7 +397,7 @@ TriangleUtils.computeVectorsDirs = function (triangle) {
             triangle.ortho_dir = dir_2.clone();
             triangle.ortho_dir.normalize();
 
-            // direction of fastest variation of weight
+            // direction of fastest constiation of weight
```
Comment (ersbeth) [inline src/utils/TriangleUtils.ts]: idem

File: src/utils/TriangleUtils.ts [hunk 52]
```diff
@@ -217,7 +429,7 @@ TriangleUtils.computeVectorsDirs = function (triangle) {
         triangle.ortho_dir.subVectors(point_iso_zero2, point_iso_zero1);
         triangle.ortho_dir.normalize();
 
-        // direction of fastest variation of weight
+        // direction of fastest constiation of weight
```
Comment (ersbeth) [inline src/utils/TriangleUtils.ts]: idem

File: src/utils/TriangleUtils.ts [hunk 53]
```diff
@@ -286,36 +498,16 @@ TriangleUtils.getParametrisedVertexAttr = function (triangle, u, v) {
...
     return triangle.v[0].getThickness() * (1 - u - v) + triangle.v[1].getThickness() * u + triangle.v[2].getThickness() * v;
-};
-
-/**
- *  @param {!Object} triangle The concerned triangle
- *  @param {number} u u coordinate
- *  @param {number} v v coordinate
- *  @return {!Material} Interpolated material
- */
-TriangleUtils.getMeanMat = function (triangle, u, v) {
```
Comment (ersbeth) [inline src/utils/TriangleUtils.ts:307]: where did that go ?

File: src/blobtree/Element.ts (hunk 1 above)
Comment (matteoJeulin) [reply src/blobtree/Element.ts:40]: Static abstract methods don't exist: https://github.com/microsoft/TypeScript/issues/34516

File: src/blobtree/areas/AreaScalisTri.ts (hunk 21 above)
Comment (matteoJeulin) [reply src/blobtree/areas/AreaScalisTri.ts:208]: It is returned after the loop

File: src/utils/TriangleUtils.ts (hunk 49 above)
Comment (matteoJeulin) [reply src/utils/TriangleUtils.ts:142]: This function is new, it is here to replace computeVectorDirs as it is more general in its typing

File: src/utils/TriangleUtils.ts (hunk 53 above)
Comment (matteoJeulin) [reply src/utils/TriangleUtils.ts:307]: This function is misplaced and not used in this repository. Maxime told me to remove it

File: src/blobtree/MaxNode.ts [hunk 54]
```diff
@@ -40,14 +40,9 @@ export class MaxNode extends Node {
                 self.addChild(c);
             });
         }
-
-        // temp consts to speed up evaluation by avoiding allocations
```
Comment (ersbeth) [inline src/blobtree/MaxNode.ts]: this comment has disappeared but you should keep it, move it above temps' declarations

File: src/blobtree/MinNode.ts [hunk 55]
```diff
@@ -41,15 +41,9 @@ export class MinNode extends Node {
                 self.addChild(c);
             });
         }
-
-        // temp consts to speed up evaluation by avoiding allocations
```
Comment (ersbeth) [inline src/blobtree/MinNode.ts]: keep comment

File: src/blobtree/RicciNode.ts [hunk 56]
```diff
@@ -40,21 +40,12 @@ export class RicciNode extends Node {
                 self.addChild(c);
             });
         }
-
-        // Tmp consts to speed up computation (no reallocations)
```
Comment (ersbeth) [inline src/blobtree/RicciNode.ts]: keep comment

File: src/blobtree/ScaleNode.ts [hunk 57]
```diff
@@ -40,14 +40,6 @@ export class ScaleNode extends Node {
                 self.addChild(c);
             });
         }
-
-        // temp consts to speed up evaluation by avoiding allocations
```
Comment (ersbeth) [inline src/blobtree/ScaleNode.ts]: keep comment

File: src/blobtree/areas/AreaScalisTri.ts [hunk 58]
```diff
@@ -27,7 +27,7 @@ export class AreaScalisTri extends Area {
     unit_normal: Vector3;
     main_dir: Vector3;
     equal_weights: boolean;
-    segParams: any; // Define type for segParams as per your actual structure
+    segParams: SegParam[]; // Define type for segParams as per your actual structure
```
Comment (ersbeth) [inline src/blobtree/areas/AreaScalisTri.ts]: You can remove the comment now

**Review by ersbeth — COMMENTED** [review]
Last changes are ok.

**Review by ersbeth — COMMENTED** [review]
Changes look good to me.

**Review by ersbeth — COMMENTED** [review]
Changes look good to me.
